File: linker/lib/instruction.rb
class Instruction
  attr_accessor :type, :address, :op_code, :word
  
  def initialize(type, word)
    @type = type
    @address = word % 1000
    @op_code = (word - @address) / 1000
  end
  
  def update_address(opts = {})
    case type
    when InstructionType::R then 
      validate_and_update_relative_address(opts[:base_address], opts[:size])
    when InstructionType::E then
      validate_and_update_external_address(opts[:symbol])
    end
    
    validate_machine_size unless type == InstructionType::I
  end
  
  def to_s
    text = word.to_s
    text += " " + errors.join(" ") if errors.any?
    text
  end
  
  def word
    (op_code * 1000) + address
  end
  
  def errors
    @errors ||= []
  end
  
  def error_message
    errors.join(" ")
  end

protected

  def validate_machine_size
    if address > Machine::SIZE
      errors << "Error: Absolute address exceeds machine size; zero used."
      @address = 0
    end
  end
  
  def validate_and_update_relative_address(base_address, size)
    if address > size
      @address = 0
      errors << "Error: Relative address exceeds module size; zero used."
    else
      @address += base_address
    end
  end
  
  def validate_and_update_external_address(symbol)
    if addr = SymbolTable.address(symbol)
      @address = addr
    else
      @address = 0
      errors << "Error: #{symbol} is not defined; zero used."
    end
  end
  
end

File: java/linker/src/lib/Instruction.java
package lib;

import java.util.ArrayList;

/**
 * User: Ross Kaffenberger
 * Date: Sep 24, 2009
 * Time: 1:02:16 AM
 */
public class Instruction {
    protected String type;
    private int address;
    private ArrayList<String> errors;
    private int opCode;
    private String symbol;

    public Instruction(String type, int word) {
        this.type = type;
        this.address = word % 1000;
        this.opCode = (word - this.address) / 1000;
        this.errors = new ArrayList<String>();
    }

    public int getAddress() {
        return address;
    }

    public void updateAddress(String symbol, int baseAddress, int size) {
        if (this.type.equals("R")) {
            validateAndUpdateRelativeAddress(baseAddress, size);
        } else if (this.type.equals("E")) {
            validateAndUpdateExternalAddress(symbol);
        }

        if (!this.type.equals("I")) {
            validateMachineSize();
        }
    }

    private void setAddress(int address) {
        this.address = address;
    }

    public String toString() {
        String text = "" + getWord();
        if (!isValid()) {
            for (String error : this.errors) {
                text += " " + error;
            }
        }
        return text;
    }

    public boolean isValid() {
        return this.errors.isEmpty();
    }

    public int getWord() {
        return (this.opCode * 1000) + this.address;
    }

    private void validateMachineSize() {
        if(getAddress() > Machine.SIZE) {
            this.errors.add("Error: Absolute address exceeds machine size; zero used");
            setAddress(0);
        }
    }

    private void validateAndUpdateRelativeAddress(int baseAddress, int size) {
        if (getAddress() > size) {
            setAddress(0);
            this.errors.add("Error: Relative address exceeds length module size; zero used");
        } else {
            setAddress(getAddress() + baseAddress);
        }
    }

    private void validateAndUpdateExternalAddress(String symbol) {
        if (symbol == null) {
            this.errors.add("Error: External address exceeds length of use list; treated as immediate");
        } else if (SymbolTable.address(symbol) != null) {
            this.symbol = symbol;
            setSymbol(symbol);
            setAddress((Integer) SymbolTable.address(symbol));
        } else {
            setAddress(0);
            this.errors.add("Error: " + symbol + " is not defined; zero used");
        }

    }

    private void setSymbol(String symbol) {
        this.symbol = symbol;
    }

    public String getType() {
        return this.type;
    }

    public String getSymbol() {
        return this.symbol;
    }
}
